sdk/src/types/block/output/feature/metadata.rs

```diff
@@ -1,24 +1,56 @@
 // Copyright 2021-2022 IOTA Stiftung
 // SPDX-License-Identifier: Apache-2.0
 
-use alloc::{boxed::Box, string::String, vec::Vec};
-use core::{ops::RangeInclusive, str::FromStr};
-
-use packable::{bounded::BoundedU16, prefix::BoxedSlicePrefix};
+use alloc::{
+    collections::BTreeMap,
+    string::{String, ToString},
+    vec::Vec,
+};
+use core::ops::{Deref, RangeInclusive};
+
+use packable::{
+    bounded::{BoundedU16, BoundedU8},
+    prefix::{BTreeMapPrefix, BoxedSlicePrefix},
+};
 
 use crate::types::block::{output::StorageScore, protocol::WorkScore, Error};
 
 pub(crate) type MetadataFeatureLength =
     BoundedU16<{ *MetadataFeature::LENGTH_RANGE.start() }, { *MetadataFeature::LENGTH_RANGE.end() }>;
 
+pub(crate) type MetadataFeatureKeyLength = BoundedU8<1, { u8::MAX }>;
+pub(crate) type MetadataFeatureValueLength = BoundedU16<0, { u16::MAX }>;
+
 /// Defines metadata, arbitrary binary data, that will be stored in the output.
 #[derive(Clone, Eq, PartialEq, Ord, PartialOrd, Hash, packable::Packable)]
-#[packable(unpack_error = Error, with = |err| Error::InvalidMetadataFeatureLength(err.into_prefix_err().into()))]
+#[packable(unpack_error = Error, with = |err| Error::InvalidMetadataFeature(err.to_string()))]
 pub struct MetadataFeature(
     // Binary data.
-    pub(crate) BoxedSlicePrefix<u8, MetadataFeatureLength>,
+    #[packable(verify_with = verify_keys_packable)]
+    pub(crate)  BTreeMapPrefix<
+        BoxedSlicePrefix<u8, MetadataFeatureKeyLength>,
+        BoxedSlicePrefix<u8, MetadataFeatureValueLength>,
+        MetadataFeatureLength,
+    >,
 );
 
+fn verify_keys_packable<const VERIFY: bool>(
+    map: &BTreeMapPrefix<
+        BoxedSlicePrefix<u8, MetadataFeatureKeyLength>,
+        BoxedSlicePrefix<u8, MetadataFeatureValueLength>,
+        MetadataFeatureLength,
+    >,
+) -> Result<(), Error> {
+    if VERIFY {
+        for key in map.keys() {
+            if !key.is_ascii() {
+                return Err(Error::NonAsciiMetadataKey(key.to_vec()));
+            }
+        }
```

Comment:
```rust
if map.keys().any(|key| !key.is_ascii()) {
    return Err(...)
}
```

Reply:
a bit more readable, rusty and less nested IMO

Reply:
Is that really all that needs to be verified? What uniqueness of the keys and the minimum number of entries count?

Reply:
It was probably done with the `for` to have access to the key

Reply:
yes, somehow overlooked that :+1: